webapp/src/components/incoming_calls/hooks.ts

```diff
@@ -0,0 +1,99 @@
+import {GlobalState} from '@mattermost/types/store';
+import {NotificationLevel} from 'mattermost-redux/constants/channels';
+import {getChannel} from 'mattermost-redux/selectors/entities/channels';
+import {getCurrentUser} from 'mattermost-redux/selectors/entities/users';
+import {useEffect} from 'react';
+import {useDispatch, useSelector, useStore} from 'react-redux';
+
+import {DID_RING_FOR_CALL} from 'src/action_types';
+import {dismissIncomingCallNotification, showSwitchCallModal} from 'src/actions';
+import {RING_LENGTH} from 'src/constants';
+import {logDebug} from 'src/log';
+import {connectedChannelID, didRingForCall} from 'src/selectors';
+import {IncomingCallNotification} from 'src/types/types';
+import {desktopGTE, getChannelURL, sendDesktopEvent, shouldRenderDesktopWidget} from 'src/utils';
+import {notificationSounds} from 'src/webapp_globals';
+
+export const useDismissJoin = (callID: string, global = false) => {
+    const store = useStore();
+    const dispatch = useDispatch();
+    const connectedID = useSelector(connectedChannelID) || '';
+
+    const onDismiss = () => {
+        dispatch(dismissIncomingCallNotification(callID));
+    };
+
+    const onJoin = () => {
+        dispatch(dismissIncomingCallNotification(callID));
+
+        if (connectedID) {
+            if (global && desktopGTE(5, 5)) {
+                logDebug('sending calls-join-request message to desktop app');
+                sendDesktopEvent('calls-join-request', {
+                    callID,
+                });
+                return;
+            }
+            if (global) {
+                logDebug('sending calls-widget-channel-link-click and calls-joined-call message to desktop app');
+                const currentChannel = getChannel(store.getState(), connectedID);
+                const channelURL = getChannelURL(store.getState(), currentChannel, currentChannel.team_id);
+                sendDesktopEvent('calls-widget-channel-link-click', {pathName: channelURL});
+                sendDesktopEvent('calls-joined-call', {
+                    type: 'calls-join-request',
+                    callID,
+                });
+                return;
+            }
+
+            dispatch(showSwitchCallModal(callID));
+            return;
+        }
+        window.postMessage({type: 'connectCall', channelID: callID}, window.origin);
+    };
+
+    return [onDismiss, onJoin];
+};
+
+export const useOnACallWithoutGlobalWidget = () => {
+    const connectedChannel = useSelector(connectedChannelID);
+    return Boolean(connectedChannel && !shouldRenderDesktopWidget());
+};
+
+export const useRinging = (call: IncomingCallNotification, onWidget: boolean) => {
+    const dispatch = useDispatch();
+    const currentUser = useSelector(getCurrentUser);
+    const callUniqueID = `${call.callID}${call.startAt}`;
```

Comment:
Ok, made the change. This was bigger than I thought it would be. And there are still some confusing usages of channelID <-> callID (recording, sending messages to the desktop).

Reply:
That's okay, as long as they are used in a way that doesn't cause conflicts we should be good.